mrmustard/physics/representations.py

```diff
@@ -0,0 +1,305 @@
+    >>> assert psi3.A.shape == (2, 1, 1)  # stacked along batch dimension
+    >>> psi4 = psi1[0] @ psi2[0]  # contract wires 0 on each (inner product)
+    >>> assert psi4.A.shape == (1,)  # A is 0x0 now (no wires left)
+    >>> psi5 = psi1 & psi2  # outer product (tensor product)
+    >>> rho = psi1.conj() & psi1   # outer product (this is now the density matrix)
+    >>> assert rho.A.shape == (1, 2, 2)  # we have two wires now
+    >>> assert np.allclose(rho.trace((0,), (1,)), np.abs(c)**2)
+
+
+    Args:
+        A: batch of quadratic coefficient :math:`A_i`
+        b: batch of linear coefficients :math:`b_i`
+        c: batch of arrays :math:`c_i` (default: [1.0])
+    """
+
+    def __init__(
+        self,
+        A: Batch[ComplexMatrix],
+        b: Batch[ComplexVector],
+        c: Batch[ComplexTensor] = [1.0],
+    ):
+        r"""Initializes the Bargmann representation. Args can be passed non-batched,
+        they will be automatically broadcasted to the correct batch shape.
+
+        Args:
+            A: batch of quadratic coefficient :math:`A_i`
+            b: batch of linear coefficients :math:`b_i`
+            c: batch of arrays :math:`c_i` (default: [1.0])
+        """
+        self.ansatz = PolyExpAnsatz(A, b, c)
+
+    def __call__(self, z: ComplexTensor) -> ComplexTensor:
+        r"""Evaluates the Bargmann function at the given array of points."""
+        return self.ansatz(z)
+
+    def from_ansatz(self, ansatz: PolyExpAnsatz) -> Bargmann:
+        r"""Returns a Bargmann object from an ansatz object."""
+        return self.__class__(ansatz.A, ansatz.b, ansatz.c)
+
+    @property
+    def A(self) -> Batch[ComplexMatrix]:
+        r"""
+        The batch of quadratic coefficient :math:`A_i`.
+        """
+        return self.ansatz.A
+
+    @property
+    def b(self) -> Batch[ComplexVector]:
+        r"""
+        The batch of linear coefficients :math:`b_i`
+        """
+        return self.ansatz.b
+
+    @property
+    def c(self) -> Batch[ComplexTensor]:
+        r"""
+        The batch of arrays :math:`c_i`.
+        """
+        return self.ansatz.c
+
+    def conj(self):
+        r"""
+        The conjugate of this Bargmann object.
+        """
+        new = self.__class__(math.conj(self.A), math.conj(self.b), math.conj(self.c))
+        new._contract_idxs = self._contract_idxs
+        return new
+
+    def __getitem__(self, idx: int | tuple[int, ...]) -> Bargmann:
+        r"""Returns a copy of self with the given indices marked for contraction."""
+        idx = (idx,) if isinstance(idx, int) else idx
+        for i in idx:
+            if i >= self.ansatz.dim:
+                raise IndexError(
+                    f"Index {i} out of bounds for ansatz {self.ansatz.__class__.__qualname__} of dimension {self.ansatz.dim}."
+                )
+        new = self.__class__(self.A, self.b, self.c)
+        new._contract_idxs = idx
+        return new
+
+    def __matmul__(self, other: Bargmann) -> Bargmann:
```

Comment:
I haven't found any check of existence of contract_idxs along the way of this function, from matmul to contract_two_Abc to join_Abc. Should we add a sanity check?

Reply:
what do you mean check of existence?